istioctl/cmd/add-to-mesh.go

```diff
@@ -0,0 +1,204 @@
+Warning: This command will modify the deployment's yaml file. So you'd better backup 
+the yaml files before hand,eg: kubectl describe deployment -l app=details > deployments.yaml
+THIS COMMAND IS STILL UNDER ACTIVE DEVELOPMENT AND NOT READY FOR PRODUCTION USE.
+`,
+		Example: `istioctl experimental add-to-mesh service productpage`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return fmt.Errorf("expecting service name")
+			}
+			client, err := interfaceFactory(kubeconfig)
+			if err != nil {
+				return err
+			}
+			var sidecarTemplate, valuesConfig string
+			ns := handlers.HandleNamespace(namespace, defaultNamespace)
+			writer := cmd.OutOrStdout()
+
+			meshConfig, err := setupParameters(&sidecarTemplate, &valuesConfig)
+			if err != nil {
+				return err
+			}
+			matchingDeployments, err := findDeploymentsForSvc(client, ns, args[0])
+			if err != nil {
+				return err
+			}
+			if len(matchingDeployments) == 0 {
+				fmt.Fprintf(writer, "No deployments found for service %s.%s", args[0], ns)
+				return nil
+			}
+			return injectSideCarIntoDeployment(client, matchingDeployments, sidecarTemplate, valuesConfig,
+				args[0], ns, meshConfig, writer)
+		},
+	}
+	return cmd
+}
+func setupParameters(sidecarTemplate, valuesConfig *string) (*meshconfig.MeshConfig, error) {
+	var meshConfig *meshconfig.MeshConfig
+	var err error
+	if meshConfigFile != "" {
+		if meshConfig, err = istiocmd.ReadMeshConfig(meshConfigFile); err != nil {
+			return nil, err
+		}
+	} else {
+		if meshConfig, err = getMeshConfigFromConfigMap(kubeconfig); err != nil {
+			return nil, err
+		}
+	}
+	if injectConfigFile != "" {
+		injectionConfig, err := ioutil.ReadFile(injectConfigFile) // nolint: vetshadow
+		if err != nil {
+			return nil, err
+		}
+		var injectConfig inject.Config
+		if err := yaml.Unmarshal(injectionConfig, &injectConfig); err != nil {
+			return nil, multierr.Append(fmt.Errorf("loading --injectConfigFile"), err)
+		}
+		*sidecarTemplate = injectConfig.Template
+	} else if *sidecarTemplate, err = getInjectConfigFromConfigMap(kubeconfig); err != nil {
+		return nil, err
+	}
+	if valuesFile != "" {
+		valuesConfigBytes, err := ioutil.ReadFile(valuesFile) // nolint: vetshadow
+		if err != nil {
+			return nil, err
+		}
+		*valuesConfig = string(valuesConfigBytes)
+	} else if *valuesConfig, err = getValuesFromConfigMap(kubeconfig); err != nil {
+		return nil, err
+	}
+	return meshConfig, err
+}
+
+func injectSideCarIntoDeployment(client kubernetes.Interface, deps []appsv1.Deployment, sidecarTemplate, valuesConfig,
+	svcName, svcNamespace string, meshConfig *meshconfig.MeshConfig, writer io.Writer) error {
+	var errs error
+	for _, dep := range deps {
+		log.Debugf("updating deployment %s.%s with Istio sidecar injected",
+			dep.Name, dep.Namespace)
+		newDep, err := inject.IntoObject(sidecarTemplate, valuesConfig, meshConfig, &dep)
```

Comment:
I tried this on a pod that was already injected and received the message

```
Skipping injection because "" has injected "istio-proxy" sidecar already
deployment details-v1.default updated successfully with Istio sidecar injected.
```

The `""` is a problem.  It does not come from your code, but is it possible to fix this message by setting the metadata.Name on `dep` before calling `IntoObject()`?

Reply:
Fixed